### ISSUE TITLE
Multithreading

### ISSUE THREAD

**Issue by zsolt777** [jchristn/WatsonTcp#93]
Hi there! I've just started to browse the code, I like it. Please let me share my thoughts that came into my mind at first glance.

WatsonTcpServer.cs
```
        private ConcurrentDictionary<string, DateTime> _ClientsLastSeen = new ConcurrentDictionary<string, DateTime>();
...
        private void UpdateClientLastSeen(string ipPort)
        {
            if (_ClientsLastSeen.ContainsKey(ipPort)) _ClientsLastSeen.TryRemove(ipPort, out DateTime ts);
            _ClientsLastSeen.TryAdd(ipPort, DateTime.Now);
        }
```
1. Why don't you use _ClientsLastSeen.AddOrUpdate instead of ContainsKey+TryRemove+TryAdd? These operations would be locked by ConcurrentDictionary internally, which is why ConcurrentDictionary exists. 
2. Or why do you use ContainsKey at all, since TryRemove also check if the key is in the dictionary, so the check is ran twice
3. Usually "Contains+Remove+Add" should be locked to ensure that the content of the dictionary does not changed by another thread while "this" thread checks it and makes decisions.

Statistics:
```
                    _Statistics.ReceivedMessages = _Statistics.ReceivedMessages + 1;
                    _Statistics.ReceivedBytes += msg.ContentLength; 
```
I think both of them can lead to bad result. ReceivedMessages and ReceivedBytes are type of LONG, so 64 bit. 
1. If you run this code on a 32 bit machine then the addition is not atomic, so the result can be a not expected number.
2. On a 64 bit machine if the variable is not aligned (so the operation cannot be atomic), then it also can lead to an unexpected number
3. The "+=" and "+1" statements are actually a "read+add+write" operations (so not atomic), so if multiple threads run this code at the same time, then ReceivedMessages would be incremented by 1 instead of 3.
4. I think Interlocked.Add, or Interlocked.Increment would be a better approach.

Thank you, cheers

**Comment by jchristn** [comment]
Hi @zsolt777 thanks for the thorough review!  It is much appreciated and I will integrate these changes ASAP.  Completely agree with ```Interlocked.Add``` and ```Interlocked.Increment``` - I've been bitten by multi-thread adding/incrementing in the past.  Will report back when these changes are made!  Cheers, Joel

**Comment by jchristn** [comment]
Hi @zsolt777 just committed and published v4.3.0.2 to NuGet.  Please re-open if this isn't along the lines of what you were recommending!  

NuGet: https://www.nuget.org/packages/WatsonTcp/4.3.0.2
Commit: https://github.com/jchristn/WatsonTcp/commit/45e6cf04e29b50eb6dcf1b5b107053d58a99aa76

Also hope you don't mind I added your handle to the README as a thank you for helping make WatsonTcp better!

**Comment by zsolt777** [comment]
Hi @jchristn , perfect! Thank you, cheers!
